src/ast/Program.h

```diff
@@ -56,21 +59,95 @@ namespace souffle::ast {
  */
 class Program : public Node {
 public:
+    // Requirements for storing top-level nodes:
+    // 1) remove-by-identity needs to be at least `O(lg n)`
+    // 2) traversal order must be stable (at least for pretty printing)
+    //
+    // This implies that top-level nodes need to be stored in a linked-set structure.
+    // Linked for stable traversal, and set for `O(lg n)` remove-by-identity.
+    //
+    // Furthermore, lookup of `Relation`, `Clause`, and `Directive` by name is a
+    // common operation.
+    // This is handled by storing `Relation`s, `Clause`s, and `Directive`s in a
+    // map keyed by the relation name. The remove-by-identity time is linear,
+    // but the size is much smaller, and this allows lookup by name w/o needing
+    // an analysis and avoids the issue of the analysis results becoming stale
+    // after common manipulations (e.g. `Clause` insertion).
+    //
+    // The downside is that the name must not be changed after the
+    // node is added to the program. This is sanchecked by assertions.
+    struct RelationInfo {
+        // A well formed program has exactly one declaration per relation.
+        // TODO:  Reject duplicate relation declarations in `SemanticChecker` instead of in the parser.
+        VecOwn<Relation> decls;
+        VecOwn<Clause> clauses;
+        VecOwn<Directive> directives;
+    };
+
+    using RelationInfoMap = std::map<QualifiedName, RelationInfo>;
+
+    RelationInfoMap& getRelationInfo() {
+        return relations;
+    }
+
+    RelationInfoMap const& getRelationInfo() const {
+        return relations;
+    }
+
+    RelationInfo* getRelationInfo(QualifiedName const& name) {
+        auto it = relations.find(name);
+        return it == relations.end() ? nullptr : &it->second;
+    }
+
+    RelationInfo const* getRelationInfo(QualifiedName const& name) const {
+        auto it = relations.find(name);
+        return it == relations.end() ? nullptr : &it->second;
+    }
+
     /** Return types */
     std::vector<Type*> getTypes() const;
 
     /** Return relations */
     std::vector<Relation*> getRelations() const;
 
+    /** Returns the first `Relation` declartion for a given name, if any */
+    Relation* getRelation(QualifiedName const&) const;
```

Comment:
Can we make utility functions for those? We had a quite convoluted AST in the past and it hurt us.

Reply:
Do you remember any short examples that you could share? Putting everything into `ast/utility/Utils.h` doesn't sound great either, and there are problems with member-func colliding with freestanding helpers. e.g. `ast::getName` collides w/ `Transformer::getName`, which forces everyone to use `ast::getName(x)` instead of `getName(x)`.